update ingress docs (#4043)
kubernetes-sigs/kind: site/content/docs/user/ingress.md

description: |-
  This guide covers setting up [ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) on a kind cluster.
---
## Compatibility：
This guide applies to [cloud-provider-kind](https://github.com/kubernetes-sigs/cloud-provider-kind) v0.9.0+. For older versions, refer to historical docs.

## Setting Up Ingress

Ingress exposes HTTP and HTTPS routes from outside the cluster to services within the cluster.

[Create a cluster](#create-cluster): There are two primary methods to direct external traffic to Services inside the cluster:
- using a [LoadBalancer]. 
- leverage KIND's `extraPortMapping` config option when creating a cluster to forward ports from the host.

> > **NOTE**: Since kind v0.9.0, the built-in cloud provider (cloud provider kind) natively supports Ingress. You do not need to deploy a third-party Ingress controller (e.g., Ingress NGINX, Contour) by default.

Reviewer: ```suggestion
> **NOTE**: Since cloud-provider-kind v0.9.0, it natively supports Ingress. You do not need to deploy a third-party Ingress controller (e.g., Ingress NGINX, Contour) to use this functionality```

For third-party ingress solutions (e.g., Ingress NGINX, Contour), please follow their official documentation.

> **NOTE**: Since kind v0.9.0, Gateway API is also natively supported (along with Ingress). You may also want to consider using [Gateway API](https://gateway-api.sigs.k8s.io/) instead of Ingress.
> See the official [Ingress migration guide](https://gateway-api.sigs.k8s.io/guides/migrating-from-ingress/) for details.

### Create Cluster

#### Option 1: LoadBalancer

> Create a kind cluster. The built-in cloud provider (cloud provider kind, enabled by default since v0.9.0) will automatically enable the LoadBalancer controller for Ingress to use.

Reviewer: i not a built-in cloud provider, you need to run it separately

> Create a kind cluster. The built-in cloud provider (cloud provider kind, enabled by default since v0.9.0) will automatically enable the LoadBalancer controller for Ingress to use.

Reviewer: ```suggestion
Create a kind cluster and run [Cloud Provider KIND] to enable the loadbalancer controller which ingress controllers will use through the loadbalancer API. Since version 0.9.0 [Cloud Provider KIND] also supports Ingress implementation```

Reviewer: remove Option 1 and Option 2, we do have only one option now

EOF
{{< /codeFromInline >}}

> If you want to run with multiple nodes, the built-in Ingress (since kind v0.9.0) will automatically use the configured PortMapping, no additional nodeSelector is required.

Author: > this is not accurate, cloud-provider-kind does not apply to the `extraPortMapping` section ... @BenTheElder for thoughts, should we leave both "roll your ingress and do this network tricks" and cloud provider kind?

I've refined this a bit and will keep both scenarios for now. Clarify that Kind v0.9.0+ natively supports Ingress/Gateway API via its built-in cloud provider, eliminating the need to deploy third-party controllers by default; correct the multi-node cluster configuration description (the built-in Ingress does not rely on extraPortMapping, while third-party controllers may require nodeSelector configuration, which is not recommended for most users); and add information on version compatibility and Gateway API.

EOF
{{< /codeFromInline >}}

For multi-node clusters with `extraPortMapping`:
- The built-in Ingress (via cloud-provider-kind, since v0.9.0) does not rely on `extraPortMapping` and works independently.
- If you explicitly use a third-party Ingress controller (not recommended for most users), you may need to configure `nodeSelector` to ensure the controller runs on the node with port mappings:
  {{< codeFromInline lang="yaml" >}}
  nodeSelector:
  kubernetes.io/hostname: "kind-control-plane"
  {{< /codeFromInline >}}

> ### Ingress NGINX

Reviewer: can we please remove this section now that the retirement is official ?

https://kubernetes.io/blog/2025/11/11/ingress-nginx-retirement/
